### PR TITLE
Plugins: Replace Rest filters with RestHandler wrapper

### PR DESCRIPTION
**Pull request by rjernst**
RestFilters are a complex way of allowing plugins to add extra code
before rest actions are executed. This change removes rest filters, and
replaces with a wrapper which a single plugin may provide.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/ActionListener.java [hunk 1]
```diff
@@ -45,7 +53,7 @@
      * @param <Response> the type of the response
      * @return a listener that listens for responses and invokes the consumer when received
      */
-    static <Response> ActionListener<Response> wrap(Consumer<Response> onResponse, Consumer<Exception> onFailure) {
+    static <Response> ActionListener<Response> wrap(CheckedConsumer<Response> onResponse, Consumer<Exception> onFailure) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/ActionListener.java]: 👍

File: core/src/main/java/org/elasticsearch/rest/RestController.java [hunk 2]
```diff
@@ -193,24 +147,24 @@ public void dispatchRequest(final RestRequest request, final RestChannel channel
                     threadContext.putHeader(key, httpHeader);
                 }
             }
-            if (filters.length == 0) {
-                executeHandler(request, channel, client);
+
+            final RestHandler rawHandler = getHandler(request);
+            final RestHandler handler = handlerWrapper.apply(rawHandler);
+
+            if (handler != null) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: should we allow null here? I think it must not return null but the raw handler instead? maybe I am missing something

File: core/src/main/java/org/elasticsearch/action/ActionModule.java [hunk 3]
```diff
@@ -338,7 +344,18 @@ public ActionModule(boolean ingestEnabled, boolean transportClient, Settings set
         autoCreateIndex = transportClient ? null : new AutoCreateIndex(settings, clusterSettings, resolver);
         destructiveOperations = new DestructiveOperations(settings, clusterSettings);
         Set<String> headers = actionPlugins.stream().flatMap(p -> p.getRestHeaders().stream()).collect(Collectors.toSet());
-        restController = new RestController(settings, headers);
+        UnaryOperator<RestHandler> restWrapper = null;
+        for (ActionPlugin plugin : actionPlugins) {
+            UnaryOperator<RestHandler> newRestWrapper = plugin.getRestHandlerWrapper(threadPool.getThreadContext());
+            if (newRestWrapper != null) {
+                logger.debug("Using REST wrapper from plugin " + plugin.getClass().getName());
+                if (restWrapper != null) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/ActionModule.java]: I am not sure if we can do this in 5.x like this. There might be users relying on the filters... The other option is to just have multple wrappers and for iterate over them? passing the last wrapped one to the next wrapper?

**Review by s1monw — COMMENTED** [review]
I love it, I added a couple of comments / questions

File: core/src/main/java/org/elasticsearch/rest/RestController.java (hunk 2 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: I have null meaning "don't use a wrapper", in which case we convert the null here to an identity wrapper to simplify the code which uses the wrapper.

File: core/src/main/java/org/elasticsearch/action/ActionModule.java (hunk 3 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/ActionModule.java]: alright, I think this is so low level we are good with 1 and only 1. I think it's saver anyway

File: core/src/main/java/org/elasticsearch/rest/RestController.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: sorry I dont' follow. if the wrapper returns null we say we didn't find a handler, are we ok with this?

File: core/src/main/java/org/elasticsearch/action/ActionModule.java (hunk 3 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/ActionModule.java]: > The notion of rest filters goes along with action filters that we also have, and in the java world it's a pretty familiar concept. I don't see how much this specific change simplifies things code-wise, but that is a different problem.

I disagree with this but we can discuss

File: core/src/main/java/org/elasticsearch/rest/RestController.java (hunk 2 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: Ok I see now. You are right, this was a bug. I pushed 65501a8.

File: core/src/main/java/org/elasticsearch/action/ActionModule.java (hunk 3 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/action/ActionModule.java]: > shall we check in the community which plugins may be using rest filters first?

Can you name a single plugin out there that implements a RestFilter? Let alone one that would work composed with another plugin implementing one? I searched for "implements RestFilter handleRequest" (the handleRequest is to distinguish from some other http libraries). All I see are forks of elasticsearch, no actual RestFilter implementations.

I'm in no rush to get this in, but I also don't think such a low level change to how plugins operate should have to wait for 6.0. If we did break such a use case, we could always implement support at that time if it was truly necessary and have in the next minor version. But I don't think we should overengineer something for what seems to me to not only be an esoteric and _theoretical_ case.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/ActionModule.java]: I have no idea if there are plugins like this out there to be honest. I just see it as an odd change that you can have only one plugin installed that can do things before executing rest actions. And, my personal opinion, I don't see having a filter chain and multiple filters with ordering as over-engineering. But again, my own personal opinion and it doesn't mean that it should block the change, especially if I am the only one seeing it this way.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/ActionModule.java]: we crossed streams, I replied before reading your comment Simon. No problem, go ahead.

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/ActionModule.java]: it's no secret that we use this for security, in such a case we really want to make sure nobody else has registered one. That is also true for anybody else that does stuff like this if you are the one protecting the handler with auth/autz you want to be sure nothing is executed before you. Now that said, I think as a compromise we could have 1 wrapper that is the primary wrapper and a set of wrappers that are unordered. Yet I still think we should try and find out if that is really needed and the only way is to wait until folks come to us with this. We can usually react pretty quickly on those things ie the compromise can come once it's a problem. I am not too concerned about this on that low level we are operation on.

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/ActionModule.java]: @javanna lemme know what you think

Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/ActionModule.java]: I am fine with one single wrapper given your explanations, I am happy that we looked for implementors but there weren't any.

**Comment by s1monw** [comment]
@elasticmachine test this please

**Review by s1monw — APPROVED** [review]
LGTM

**Pull request merged**